### PR TITLE
Migrate sample's DateOfBirth field to AgeDateOfBirthField

### PR DESCRIPTION
**Pull request by xispa**
## Description of the issue/feature this PR addresses

> **Warning**
> Merge https://github.com/senaite/senaite.core/pull/2310 and https://github.com/senaite/senaite.core/pull/2311 first

This Pull Request makes `senaite.patient` compatible with https://github.com/senaite/senaite.core/pull/2307 and https://github.com/senaite/senaite.core/pull/2311 by replacing the `DateTimeField` type for sample's `DateOfBirth` to a new field `AgeDateOfBirthField`, fully supported by the `AgeDoBWidget`.

## Current behavior before PR

The "DateOfBirth" field was only storing the datetime, while other attributes required for the field and widget to work properly were stored as independent attributes in the instance itself.

## Desired behavior after PR is merged

The "DateOfBirth" field stores a tuple (`dob`, `from_age`, `estimated`).

--
I confirm I have tested this PR thoroughly and coded it according to [PEP8][1]
and [Plone's Python styleguide][2] standards.

[1]: https://www.python.org/dev/peps/pep-0008
[2]: https://docs.plone.org/develop/styleguide/python.html

### CODE REVIEW

File: src/senaite/patient/api.py [hunk 1]
```diff
@@ -202,67 +205,80 @@ def to_datetime(date_value, default=None, tzinfo=None):
...
         return api.to_int(num[0], default=0)
 
-    # Extract the periods
-    years = extract_period(age_ymd, "y")
-    months = extract_period(age_ymd, "m")
-    days = extract_period(age_ymd, "d")
+    years = extract_period(ymd, "y")
+    months = extract_period(ymd, "m")
+    days = extract_period(ymd, "d")
     if not any([years, months, days]):
```
Comment (ramonski) [inline src/senaite/patient/api.py:241]: Why can't we just return `(0, 0, 0)`?
E.g. for a new born baby with 2 hours.

File: src/senaite/patient/api.py [hunk 2]
```diff
@@ -202,67 +205,80 @@ def to_datetime(date_value, default=None, tzinfo=None):
...
+def get_birth_date(age_ymd, on_date=None, default=_marker):
+    """Returns the birth date given an age in ymd format and the date when age
+    was recorded or current datetime if None
+    """
+    try:
+        ymd = to_ymd(age_ymd)
+        years, months, days = get_years_months_days(ymd)
+    except AttributeError:
+        if default is _marker:
+            raise AttributeError("No valid ymd: {}".format(age_ymd))
```
Comment (ramonski) [inline src/senaite/patient/api.py:258]: Phew, catch an `AttributeError` (which is actually none) to raise an `AttributeError` again to (probably) bypass the `IndexError` that might happen here `years, months, days = get_years_months_days(ymd)` ...

I would maybe completely skip the handling here and leave it to the `get_birth_date` to either raise a `TypeError` or return `(0, 0, 0)`.

File: src/senaite/patient/api.py [hunk 3]
```diff
@@ -202,67 +205,80 @@ def to_datetime(date_value, default=None, tzinfo=None):
...
 def get_age_ymd(birth_date, on_date=None):
     """Returns the age at on_date if not None. Otherwise, current age
     """
-    delta = get_relative_delta(birth_date, on_date)
-    return to_ymd(delta)
+    try:
+        delta = dtime.get_relative_delta(birth_date, on_date)
+        return to_ymd(delta)
+    except (ValueError, TypeError):
+        return None
```
Comment (ramonski) [inline src/senaite/patient/api.py:273]: You do everywhere the `default` handling except of here you return `None`? What happened? 😅

Maybe you can skip here the `ValueError`. At least I see only the explicit `TypeError` that is raised by `to_ymd`.

File: src/senaite/patient/browser/widgets/agedob.py [hunk 4]
```diff
@@ -19,113 +19,78 @@ class AgeDoBWidget(DateTimeWidget):
...
+            # Age entered
             ymd = map(lambda p: value.get(p), ["years", "months", "days"])
             if not any(ymd):
                 # No values set
-                return None
+                return None, {}
 
             # Age in ymd format
             ymd = filter(lambda p: p[0], zip(ymd, 'ymd'))
             ymd = "".join(map(lambda p: "".join(p), ymd))
```
Comment (ramonski) [inline src/senaite/patient/browser/widgets/agedob.py:64]: Ok, once again this dance. If it is really required, I would somehow encapsulate it in a `format_ymd` function with a good docstring / comments.

File: src/senaite/patient/content/fields.py [hunk 5]
```diff
@@ -118,3 +120,92 @@ def get_fullname(self, instance):
...
+        elif patient_api.is_ymd(value):
+            dob = patient_api.get_birth_date(value)
+            from_age = estimated = True
+
+        # store the tuple or default
+        val = (dob, from_age, estimated) if dob else self.getDefault(instance)
+        super(AgeDateOfBirthField, self).set(instance, val)
+
+    def get_date_of_birth(self, instance):
+        """Returns whether the date of birth
```
Comment (ramonski) [inline src/senaite/patient/content/fields.py:181]: incomplete/confusing docstring

File: src/senaite/patient/api.py [hunk 6]
```diff
@@ -202,67 +205,80 @@ def to_datetime(date_value, default=None, tzinfo=None):
...
+            return val
+        if default is _marker:
+            raise TypeError("delta parameter must be a relative_delta or ymd")
+        return default
 
     ymd = list("ymd")
-    diff = map(str, (delta.years, delta.months, delta.days))
+    diff = map(str, (val.years, val.months, val.days))
     age = filter(lambda it: int(it[0]), zip(diff, ymd))
     return " ".join(map("".join, age))
```
[the remark this answers is not in the document]
Comment (ramonski) [reply src/senaite/patient/api.py:221]: Ok, I've seen it in the doctest now, it generates something like `43y 2m 3d`.
As proposed further down, please encapsulate it in an own function, e.g. `format_ymd` to do this.

File: src/senaite/patient/api.py [hunk 7]
```diff
@@ -186,6 +188,7 @@ def update_patient(patient, **values):
     patient.reindexObject()
 
 
+@deprecated("Us senaite.core.api.dtime.to_dt instead")
```
[the remark this answers is not in the document]
Comment (xispa) [reply src/senaite/patient/api.py]: resolved with c06ae2ba4b0e863de7302051d4a7c705bed046ea

File: src/senaite/patient/api.py (hunk 1 above)
Comment (xispa) [reply src/senaite/patient/api.py]: yes, with latest changes it always returns `(0, 0, 0)` unless format or type are not valid

File: src/senaite/patient/api.py [hunk 8]
```diff
@@ -202,67 +205,80 @@ def to_datetime(date_value, default=None, tzinfo=None):
...
-    valid = map(lambda p: p in ymd, "ymd")
-    return any(valid)
+    values = get_years_months_days(ymd, default=None)
+    return values is not None
 
 
-def get_birth_date(age_ymd, on_date=None):
-    """Returns the birth date given an age in ymd format and the date when age
-    was recorded or current datetime if None
+def get_years_months_days(ymd, default=_marker):
```
[the remark this answers is not in the document]
Comment (xispa) [reply src/senaite/patient/api.py]: removed

File: src/senaite/patient/api.py (hunk 2 above)
Comment (xispa) [reply src/senaite/patient/api.py]: I simplified the whole thing and now it raises `TypeError` or `ValueError` only if the format is not correct

File: src/senaite/patient/api.py (hunk 3 above)
Comment (xispa) [reply src/senaite/patient/api.py:334]: Done

File: src/senaite/patient/content/fields.py (hunk 5 above)
Comment (xispa) [reply src/senaite/patient/content/fields.py:181]: fixed

File: src/senaite/patient/browser/widgets/agedob.py (hunk 4 above)
Comment (xispa) [reply src/senaite/patient/browser/widgets/agedob.py:64]: Done with e77df12688a1c7590d380afaf7ef165110d47d35 . Is now using to `to_ymd` function

**Comment by xispa** [comment]
> Phew, what a brain excursion... I left some comments on my journey through the code.

Hehe I think so. Apologies for these excursions

File: src/senaite/patient/tests/doctests/API.rst [hunk 9]
```diff
@@ -0,0 +1,202 @@
...
+
+
+Get the age
+...........
+
+Having a birth date, we can get the age at a given date:
+
+    >>> dob = dtime.to_dt("19791207")
+    >>> api.get_age_ymd(dob, on_date="20230518")
+    '43y 5m 11d'
```
Comment (ramonski) [inline src/senaite/patient/tests/doctests/API.rst:196]: Happy birthday:)

**Review by ramonski — APPROVED** [review]

**Pull request merged**